shell/platform/linux/fl_engine.cc

```diff
@@ -0,0 +1,182 @@
+
+#include "flutter/shell/platform/embedder/embedder.h"
+
+/**
+ * FlEngine:
+ *
+ * #FlEngine is an abstract class that controls a FlutterEngine.
+ */
+
+typedef struct {
+  FlDartProject* project;
+  FLUTTER_API_SYMBOL(FlutterEngine) engine;
+} FlEnginePrivate;
+
+enum { PROP_FLUTTER_PROJECT = 1, PROP_LAST };
+
+G_DEFINE_TYPE_WITH_PRIVATE(FlEngine, fl_engine, G_TYPE_OBJECT)
+
+static void* fl_engine_gl_proc_resolver(void* user_data, const char* name) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+
+  return FL_ENGINE_GET_CLASS(self)->get_proc_address(self, name);
+}
+
+static bool fl_engine_gl_make_current(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+
+  FL_ENGINE_GET_CLASS(self)->make_current(self);
+
+  return true;
+}
+
+static bool fl_engine_gl_clear_current(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+
+  FL_ENGINE_GET_CLASS(self)->clear_current(self);
+
+  return true;
+}
+
+static uint32_t fl_engine_gl_fbo_callback(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+
+  return FL_ENGINE_GET_CLASS(self)->get_fbo(self);
+}
+
+static bool fl_engine_gl_present(void* user_data) {
+  FlEngine* self = static_cast<FlEngine*>(user_data);
+
+  FL_ENGINE_GET_CLASS(self)->present(self);
+
+  return true;
+}
+
+static void fl_engine_set_property(GObject* object,
+                                   guint prop_id,
+                                   const GValue* value,
+                                   GParamSpec* pspec) {
+  FlEngine* self = FL_ENGINE(object);
+  FlEnginePrivate* priv = static_cast<FlEnginePrivate*>(fl_engine_get_instance_private(self));
+
+  switch (prop_id) {
+    case PROP_FLUTTER_PROJECT:
+      g_set_object(&priv->project,
+                   static_cast<FlDartProject*>(g_value_get_object(value)));
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+static void fl_engine_get_property(GObject* object,
+                                 guint prop_id,
+                                 GValue* value,
+                                 GParamSpec* pspec) {
+  FlEngine* self = FL_ENGINE(object);
+  FlEnginePrivate* priv = static_cast<FlEnginePrivate*>(fl_engine_get_instance_private(self));
+
+  switch (prop_id) {
+    case PROP_FLUTTER_PROJECT:
+      g_value_set_object(value, priv->project);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+static void fl_engine_dispose(GObject* object) {
+  FlEngine* self = FL_ENGINE(object);
+  FlEnginePrivate* priv = static_cast<FlEnginePrivate*>(fl_engine_get_instance_private(self));
+
+  FlutterEngineDeinitialize(priv->engine);
+  FlutterEngineShutdown(priv->engine);
+
+  g_clear_object(&priv->project);
+
+  G_OBJECT_CLASS(fl_engine_parent_class)->dispose(object);
+}
+
+static gboolean fl_engine_real_start(FlEngine* self, GError **error) {
```

Comment:
What does "real" indicate here? I'm not sure what it's in contrast to. (Is this a GObject convention?)

Reply:
Yes, you're getting a non-requested world tour of GObject conventions :) This is the default implementation of the virtual method. The 'real' is added because 'fl_engine_start' is the public function that calls the method, and you need add something to the name for the internal implementation, because C.